macrosynergy/management/types/qdf/methods.py

```diff
@@ -0,0 +1,270 @@
+        df = df.loc[df["real_date"] >= pd.to_datetime(start)]
+    if end is not None:
+        df = df.loc[df["real_date"] <= pd.to_datetime(end)]
+
+    if blacklist is not None:
+        df = apply_blacklist(df, blacklist)
+
+    if cids is None:
+        cids = sorted(df["cid"].unique())
+    else:
+        cids_in_df = df["cid"].unique()
+        cids = sorted(c for c in cids if c in cids_in_df)
+
+    if xcats is None:
+        xcats = sorted(df["xcat"].unique())
+    else:
+        xcats_in_df = df["xcat"].unique()
+        xcats = sorted(x for x in xcats if x in xcats_in_df)
+
+    if intersect:
+        cids_in_df = set.intersection(
+            *(set(df[df["xcat"] == xcat]["cid"].unique()) for xcat in xcats)
+        )
+    else:
+        cids_in_df = df["cid"].unique()
+        cids = sorted(c for c in cids if c in cids_in_df)
+
+    df = df[df["xcat"].isin(xcats)]
+    df = df[df["cid"].isin(cids)]
+
+    xcats_found = sorted(set(df["xcat"].unique()))
+    cids_found = sorted(set(df["cid"].unique()))
+    df = df.drop_duplicates().reset_index(drop=True)
+    if out_all:
+        return df, xcats_found, cids_found
+    else:
+        return df
+
+
+def update_df(
+    df: QuantamentalDataFrameBase,
+    df_add: QuantamentalDataFrameBase,
+    xcat_replace: bool = False,
+) -> QuantamentalDataFrameBase:
+    """
+    Append a standard DataFrame to a standard base DataFrame with ticker replacement on
+    the intersection.
+    """
+    if not isinstance(df, QuantamentalDataFrameBase):
+        raise TypeError("`df` must be a QuantamentalDataFrame.")
+    if not isinstance(df_add, QuantamentalDataFrameBase):
+        raise TypeError("`df_add` must be a QuantamentalDataFrame.")
+    if not isinstance(xcat_replace, bool):
+        raise TypeError("`xcat_replace` must be a boolean.")
+
+    if xcat_replace:
+        df = update_categories(df=df, df_add=df_add)
+    else:
+        df = update_tickers(df=df, df_add=df_add)
+    _sortorder = QuantamentalDataFrameBase.IndexColsSortOrder
+    return df.sort_values(_sortorder).reset_index(drop=True)
+
+
+def update_tickers(
+    df: pd.DataFrame,
+    df_add: pd.DataFrame,
+):
+    """
+    Method used to update aggregate DataFrame on a ticker level.
+    """
+    df = pd.concat([df, df_add], axis=0, ignore_index=True)
+    df = df.drop_duplicates(
+        subset=QuantamentalDataFrameBase.IndexCols,
+        keep="last",
+    ).reset_index(drop=True)
+    return df
+
+
+def update_categories(
+    df: QuantamentalDataFrameBase,
+    df_add: QuantamentalDataFrameBase,
+) -> QuantamentalDataFrameBase:
+    """
+    Method used to update the DataFrame on the category level.
+    """
+    if not isinstance(df, QuantamentalDataFrameBase):
+        raise TypeError("`df` must be a QuantamentalDataFrame.")
+    if not isinstance(df_add, QuantamentalDataFrameBase):
+        raise TypeError("`df_add` must be a QuantamentalDataFrame.")
+
+    incumbent_categories = list(df["xcat"].unique())
```

Comment:
I think you could just use `set(df["xcat"].unique())` here and then not use the set later on